### ISSUE TITLE
Fetching value of Variable unnecessarily slow

### ISSUE THREAD

**Issue by yaroslavvb** [tensorflow/tensorflow#17246]
Doing sess.run(var) is about 5x slower than sess.run(var+1). 

python [variable_fetch_bug_report.py](https://github.com/diux-dev/cluster/blob/26f8e01bd79e49fe2d1dac342dd90493f693b85c/yuxin_numpy/variable_fetch_bug_report.py)

100MB variable
fetch_cpu_variable  : 2.5 GB/sec, min: 40.74, median: 41.33, mean: 42.08
fetch_cpu_variable_add: 12.6 GB/sec, min: 7.96, median: 8.54, mean: 8.71
fetch_cpu_variable_concat: 14.0 GB/sec, min: 7.12, median: 8.14, mean: 8.28

TensorFlow version info:
version: 1.7.0-dev20180221
__git_version__: v1.6.0-rc1-337-gd100729
https://github.com/tensorflow/tensorflow/commit/d100729

**Comment by angersson** [comment]
@mrry Any clue what's going on here?

**Comment by mrry** [comment]
Presumably it's copying when it doesn't (?) need to. @alextp added the optimizations for the fetch path, so might know what could be going on here.

**Comment by alextp** [comment]
Does the problem also happen with resource variables? (tfe.Variable or tf.get_variable(..., use_resource=True)

**Comment by yaroslavvb** [comment]
Yes, same speed is with resource variables. BTW, I can get the same fast speed if I turn off all optimizers and fetch `var+0`

[tf_numpy_benchmark.py](https://github.com/diux-dev/cluster/blob/37d069c20fae6aeac10a53e3f801d29aebc5d6b4/yuxin_numpy/tf_numpy_benchmark.py)

```
python tf_numpy_benchmark.py --benchmark=fetch_cpu_variable_add --size-mb=1024 --num-iters=31
fetch_cpu_variable_add        :  21.0 GB/sec, min: 48.83, median: 56.08, mean: 55.69

python tf_numpy_benchmark.py --benchmark=fetch_cpu_resource_variable --num-iters=11 --size-mb=1024
fetch_cpu_variable            :   2.6 GB/sec, min: 401.36, median: 404.02, mean: 403.63
```

I suspect that fetching variable triggers a single threaded memcpy. Meanwhile fetching "var+0" uses multiple cores, so it's essentially a multi-threaded memory copy

**Comment by alextp** [comment]
The code which fetches tensors is https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/lib/core/ndarray_tensor.cc#L331 which calls https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.cc#L227 which triggers a copy if the refcount is not 1 (and it's never 1 for variables). It does the copy with normal memcpy. 

I remember people caring passionately about us not simply forwarding the memory when tensorflow still holds a reference to it as it can break some py_func use cases and some multithreaded use cases.

So I guess we should use a faster memcpy?

**Comment by yaroslavvb** [comment]
Yes, faster memcpy would also resolve https://github.com/tensorflow/tensorflow/issues/17233 , until there are better tools to create 64-byte aligned numpy arrays.

(PS: I wonder if this alignment requirement is moot in a first place. In PyTorch, I can inititialize tensors from unaligned numpy arrays, with memory reuse, and add them at 20 GB/second on CPU, there's no benefit in starting with aligned memory.)

**Comment by yaroslavvb** [comment]
Having, faster memcpy in TF also could be later worked into open-source distributed TensorFlow. Right now sending messages locally in Open-Source distributed TF is done at speed of single-threaded memcpy, so TF process can't take advantage of faster network cards (ie, AWS instances have 25 Gbps)

**Comment by yaroslavvb** [comment]
BTW, here's an example of multi-threaded memcpy

https://github.com/diux-dev/cluster/blob/master/psbench/memcpy.cc

**Comment by yaroslavvb** [comment]
On dual XeonV4

```
wget https://raw.githubusercontent.com/diux-dev/cluster/master/psbench/memcpy.cc
g++ -std=c++0x memcpy.cc -pthread -march=native -O6
./a.out 1000 32

Stream copy 32 threads: 4.8 ms, 20.98 GB/sec
Stream copy 32 threads: 4.9 ms, 20.53 GB/sec
Stream copy 32 threads: 4.8 ms, 20.87 GB/sec
Stream copy 32 threads: 4.9 ms, 20.36 GB/sec
Stream copy 32 threads: 4.8 ms, 20.74 GB/sec
Stream copy 32 threads: 4.7 ms, 21.39 GB/sec

```

**Comment by ekelsen** [comment]
A single threaded call to the system memcpy should be able to hit the memory bandwidth of the system except in two cases:

1) there are many small copies (not the case here)

2) A single large copy will normally hit the memcpy path that avoids polluting the L2 cache (the copies will done using streaming memory instructions), this is known to decrease absolute performance but avoids cache pollution on multiprocessor machines.  This is hardcoded behavior of glibc.  The latest versions of glibc should expose this parameter as a tunable value (I put them there, see https://www.gnu.org/software/libc/manual/html_node/Hardware-Capability-Tunables.html#Hardware-Capability-Tunables).   Can you try making the non-temporal threshold larger than your copy size?

I suspect what's happening is that when you shard the copy 32 times it drops below the hard-coded non temporal threshold and you're hitting the faster path (but blowing out the caches).

**Comment by yaroslavvb** [comment]
That benchmark was done using a loop with calls to `_mm256_stream_load_si256` and `_mm256_stream_si256` rather than memcpy. 

I tried using [regular memcpy](https://github.com/diux-dev/cluster/blob/master/psbench/memcpy_classic.cc) and the performance is very close, so I think stock memcpy is already using stream version.

However, single threaded performance is quite far from multi-threaded performance. Here's an experiment on AWS c5.18xlarge instance (dual 18-core Skylake), it runs 6x faster than single-threaded memcpy, copying 100MB array in 2.7ms. Skylakes [supposedly have](https://www.anandtech.com/show/11544/intel-skylake-ep-vs-amd-epyc-7000-cpu-battle-of-the-decade/12) 200 GB/second memory bandwidth

```
wget -N https://raw.githubusercontent.com/diux-dev/cluster/master/psbench/memcpy_fast.cc
g++ -std=c++0x memcpy_fast.cc -pthread -march=native -O6 -o memcpy_fast
numactl --cpunodebind 0 --membind 0 ./memcpy_fast 100 16
Stream copy 16 threads: 2.7 ms, 37.33 GB/sec
Stream copy 16 threads: 2.7 ms, 37.41 GB/sec
Stream copy 16 threads: 2.7 ms, 37.19 GB/sec
Stream copy 16 threads: 2.7 ms, 37.27 GB/sec
```

This [article](http://web.archive.org/web/20131223174037/http://software.intel.com/en-us/articles/memcpy-performance/) talks limitations of default `memcpy`. Since it's precompiled, it doesn't use the latest instructions (ie, Skylakes have AVX512). Apparently `icc` replaces `memcpy` with machine optimized version

**Comment by yaroslavvb** [comment]
That tunable didn't seem to make any difference for me, maybe glibc is too old (version 2.23)

```
export GLIBC_TUNABLES=glibc.tune.x86_non_temporal_threshold=100000000000
wget -N https://raw.githubusercontent.com/diux-dev/cluster/master/psbench/memcpy_classic.cc
g++ -std=c++0x memcpy_classic.cc -pthread -march=native -O6 -o memcpy_classic
./memcpy_classic 100 
memcpy: 23.1 ms, 4.34 GB/sec
memcpy: 23.1 ms, 4.34 GB/sec
memcpy: 23.0 ms, 4.34 GB/sec
memcpy: 23.0 ms, 4.34 GB/sec
```

**Comment by ekelsen** [comment]
Yes that's before it went in; 2.25 or 2.26 I believe.

That article is nearly a decade old and out of date.  The memcpy maintainer in glibc works for Intel.  There are assembly routines for every specialized architecture variant including AVX 512 in the latest versions (but AVX2 is preferred because of the large down clocks associated with using AVX 512).

I think your bandwidth calculation is off by a factor of 2, it only includes bytes read.

That being said, it does seem like I was incorrect and multi-threading does provide a benefit, especially on skylake.  However, 2.5 GB/sec from single threaded system memcpy seems way too low (even lowly skylake should get 10 GB/sec), so I don't know that we completely understand what's going on here.

**Comment by yaroslavvb** [comment]
I see.

To summarize, currently tensorflow takes 40ms to fetch a 100MB variable on CPU, whereas copying 100MB on same machine is possible to do in 2.7ms.

This delay is an issue when integrating TensorFlow with other systems. For instance resnet-50 backprop is 120ms, if you use something like Ray to synchronize parameter values between machines, this extra 40ms delay is significant

**Comment by yaroslavvb** [comment]
cc @tatianashp

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ezhulenev** [comment]
I believe that https://github.com/tensorflow/tensorflow/commit/879fc3440495d9388754cb7d1878caf034d03d61 should solve this issue (or at least make it slightly better). Though solution is CPU/platform/glibc dependent, on my CPU (Intel Broadwell with HyperThreading (6 cores) dL1:32KB dL2:256KB dL3:15MB) I see ~1.7x improvement.

**Comment by yaroslavvb** [comment]
Weird, is memmove doing multi-threaded copy? 

BTW, here's the code that does 100MB copy in 2.7 ms on Skylake. Besides multi-threading, it uses stream instructions to turn off cache, that just slows things down on large copies

```
wget -N https://raw.githubusercontent.com/diux-dev/cluster/master/psbench/memcpy_fast.cc
g++ -std=c++0x memcpy_fast.cc -pthread -march=native -O6 -o memcpy_fast
numactl --cpunodebind 0 --membind 0 ./memcpy_fast 100 16
Stream copy 16 threads: 2.7 ms, 37.33 GB/sec
Stream copy 16 threads: 2.7 ms, 37.41 GB/sec
```

**Comment by ezhulenev** [comment]
Nope, they both are single threaded, but memmove is using sse3 instruction to copy memory, while memcpy is using sse2 (that's what I see in perf). This "fix" is more like a ugly workaround for the old glibc versions.

There is a discussion in Eigen (https://bitbucket.org/eigen/eigen/pull-requests/292/adds-a-fast-memcpy-function-to-eigen/diff), but that change was later rolled back because it's not _always_ guaranteed to be faster.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by alextp** [comment]
I'm closing this issue now since I think @ezhulenev 's change fixed the problem. Please reopen if that's not the case.

**Comment by ppwwyyxx** [comment]
From the discussion in the eigen thread (by @rmlarsen ), looks like the above fix is only a workaround for suckiness in certain version of glibc (where memmove is even faster than memcpy).

**Comment by yaroslavvb** [comment]
The change improves single threaded copy, but it's still much faster to fetch a+0 instead of a, because the former does multi-threaded

**Comment by yaroslavvb** [comment]
BTW, I just did a benchmark with TF on DLAMI v11 (with MKL) on Skylake 18 core, and I'm seeing 10x improvement if I fetch "var+0" instead of var. Could try with nightly because of https://github.com/tensorflow/tensorflow/issues/20887

# fetching variable is slow from TF, because its single threaded
python tf_numpy_benchmark.py --benchmark=fetch_cpu_variable --num-iters=200
fetch_cpu_variable            :   2.0 GB/sec, min: 50.75, median: 50.89, mean: 50.94

# however, there's a trick, adding 0 to variable makes it multithreaded 10x faster
python tf_numpy_benchmark.py --benchmark=fetch_cpu_variable_plus0 --num-iters=200
fetch_cpu_variable            :  33.1 GB/sec, min:  3.02, median:  3.21, mean:  3.22

**Comment by yaroslavvb** [comment]
Still present in tf_nightly ("b'v1.9.0-rc2-572-geadcdf91aa'")

**Comment by ezhulenev** [comment]
@yaroslavvb Mkl has it's own kernel for Add, if it's easy for you to do the same test with default Tensorflow, I'd be very interested to know if it's any significant difference between MKL and Eigen for such simple kernel

**Comment by yaroslavvb** [comment]
Using stock tfnightly I get about 8x speed-up instead of 10x by using +0 trick (https://github.com/tensorflow/tensorflow/commit/eadcdf91aa)

# regular fetching
python tf_numpy_benchmark.py --benchmark=fetch_cpu_variable
fetch_cpu_variable            :   2.0 GB/sec, min: 49.50, median: 49.62, mean: 49.68

# stock tensorflow
python tf_numpy_benchmark.py --benchmark=fetch_cpu_variable_plus0
fetch_cpu_variable_plus0      :  16.2 GB/sec, min:  6.19, median:  6.94, mean:  6.94

# DLAMI version
python tf_numpy_benchmark.py --benchmark=fetch_cpu_variable_plus0
fetch_cpu_variable_plus0      :  25.8 GB/sec, min:  3.87, median:  4.03, mean:  4.04

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 119 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @alextp: It has been 134 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jvishnuvardhan** [comment]
@yaroslavvb Please let us know if this this still an issue with latest TF or can we close this issue? Thanks!

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity. Please open a new ticket when new information becomes available. Thanks!
